pkg/rollout/rollout.go

```diff
@@ -97,35 +103,64 @@ func (r *Rollout) UpdateService(svc *run.Service) (*run.Service, error) {
 		return nil, nil
 	}
 	r.log.Debugf("%q is the stable revision", stable)
+	r.log = r.log.WithField("stable", stable)
 
 	candidate := DetectCandidateRevisionName(svc, stable)
 	if candidate == "" {
 		r.log.Info("could not determine candidate revision")
 		return nil, nil
 	}
-	r.log.Debugf("%q is the candidate revision", candidate)
+	r.log.Debugf("%q is the candidate revision", stable)
+	r.log = r.log.WithField("candidate", candidate)
+
+	// A new candidate does not have metrics yet, so it can't be diagnosed.
+	if isNewCandidate(svc, candidate) {
+		r.log.Debug("new candidate found")
+		svc = r.RollForward(svc, stable, candidate)
```

Comment:
In my initial read of RollForward and Rollback, I assumed these functions included the service update and wondered at the lack of error handling. Since they are not responsible for the service update, maybe it should be called PrepareRollForward/PrepareRollback.